extensions/amp-next-page/1.0/service.js

```diff
@@ -237,26 +246,77 @@ export class NextPageService {
+   * @param {number=} forgetPageCountForTesting
+   * @return {!Promise}
+   * @private
    */
-  hidePreviousPages(index) {
+  hidePreviousPages_(index, forgetPageCountForTesting) {
+    // The distance (in pages) to the currently visible page after which
+    // we start unloading pages from memory
+    const forgetPageCount =
+      forgetPageCountForTesting === undefined
+        ? FORGET_PAGE_COUNT
+        : forgetPageCountForTesting;
+
+    const scrollingDown = this.scrollDirection_ === Direction.DOWN;
+    // Hide the host (first) page if needed
+    if (scrollingDown && this.hostPage_.isVisible()) {
+      this.hostPage_.setVisibility(VisibilityState.HIDDEN);
+    }
+
     // Get all the pages that the user scrolled past (or didn't see yet)
-    const previousPages =
-      this.scrollDirection_ === Direction.UP
-        ? this.pages_.slice(index + 1)
-        : this.pages_.slice(0, index);
+    const previousPages = scrollingDown
+      ? this.pages_.slice(1, index).reverse()
+      : this.pages_.slice(index + 1);
 
     // Find the ones that should be hidden (no longer inside the viewport)
-    previousPages
-      .filter(page => {
-        const shouldHide =
-          page.relativePos === ViewportRelativePos.LEAVING_VIEWPORT ||
-          page.relativePos === ViewportRelativePos.OUTSIDE_VIEWPORT ||
-          page === this.hostPage_;
-        return shouldHide && page.isVisible();
-      })
-      .forEach(page => page.setVisibility(VisibilityState.HIDDEN));
+    return Promise.all(
+      previousPages
+        .filter(page => {
+          const shouldHide =
+            page.relativePos === ViewportRelativePos.LEAVING_VIEWPORT ||
+            page.relativePos === ViewportRelativePos.OUTSIDE_VIEWPORT;
+          return shouldHide;
+        })
+        .map((page, away) => {
+          // Hide all pages that are in the viewport
+          if (page.isVisible()) {
+            page.setVisibility(VisibilityState.HIDDEN);
+          }
+          // Pause those that are too far away
+          if (away >= forgetPageCount) {
+            return page.pause();
+          }
+        })
+    );
+  }
+
+  /**
+   * Makes sure that all pages that are a few pages away from the
+   * currently visible page are re-inserted (if forgotten) and
+   * ready to become visible soon
+   * @param {number} index index of the page to start at
+   * @param {number=} forgetPageCountForTesting
+   * @private
+   */
+  resumeForgottenPages_(index, forgetPageCountForTesting) {
+    // The distance (in pages) to the currently visible page after which
+    // we start unloading pages from memory
+    const forgetPageCount =
+      forgetPageCountForTesting === undefined
+        ? FORGET_PAGE_COUNT
+        : forgetPageCountForTesting;
+
+    // Get all the pages that should be resumed
+    const nearViewportPages = this.pages_
+      .slice(1) // Ignore host page
+      .slice(index - forgetPageCount - 1, index + forgetPageCount + 1)
```

Comment:
If `this.pages_.length` is  `10` and the index here is `4`, the second `slice` here calls `(-2, 10)` after removing the first page. You should add unit tests for these negative value calls to confirm this is the right behavior.

Also, how was page count of `5` decided?

Reply:
Page count of 5 is based on manual tests but will need UX review before launch.